### PR TITLE
RELATED: RAIL-417 executeAfm

### PR DESCRIPTION
**Pull request by zbycz**
ready-to-be-merged

//update: Latest release `gooddata@4.3.1-pavelzgdc-pzb-execute-afm-2017-11-20T10-36-56-807Z`

### CODE REVIEW

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-js-opensource-repository-check/51/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-el7/292/ : SUCCESS in 1m 57s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-docker-el7/3/ : SUCCESS in 2m 45s (non-voting)

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-js-opensource-repository-check/52/ : SUCCESS in 10s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-el7/293/ : SUCCESS in 1m 26s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-docker-el7/4/ : SUCCESS in 2m 06s (non-voting)

File: src/execute-afm.js [hunk 1]
```diff
@@ -0,0 +1,83 @@
+// Copyright (C) 2007-2014, GoodData(R) Corporation. All rights reserved.
+import { get, mapValues, clone } from 'lodash';
+import invariant from 'invariant';
+import { ajax, post, parseJSON } from './xhr';
+import { queryString } from './util';
+
+const PAGE_SIZE = 500;
+
+function getDimensionality(execution) {
+    return get(execution, 'execution.resultSpec.dimensions', [1, 2]).length;
```
Comment (zbycz) [inline src/execute-afm.js]: Is it ok to default to "2 dimensions"?

File: src/execute-afm.js [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    const rowOffset = offset[0];
+    if (resultSoFar.data[rowOffset]) { // appending to existing rows
+        for (let i = 0; i < data.length; i += 1) {
+            resultSoFar.data[i + rowOffset].push(...data[i]);
+        }
+    } else {
+        resultSoFar.data.push(...data);
+    }
+    return resultSoFar;
+}
```
Comment (zbycz) [inline src/execute-afm.js]: This works only for one or two dimensions. Is it ok to leave `invariant` only in the "main" function?

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-js-opensource-repository-check/53/ : SUCCESS in 9s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-el7/294/ : SUCCESS in 2m 06s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-docker-el7/5/ : SUCCESS in 2m 11s (non-voting)

File: test/execution/execution-afm.test.js [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+    });
+    it('should work for 2 dimensions', () => {
+        expect(nextPageOffset({ offset: [0, 0], total: [501, 501] })).toEqual([0, 500]);
+        expect(nextPageOffset({ offset: [0, 500], total: [501, 501] })).toEqual([500, 0]);
+        expect(nextPageOffset({ offset: [500, 0], total: [501, 501] })).toEqual([500, 500]);
+        expect(nextPageOffset({ offset: [500, 500], total: [501, 501] })).toEqual(false);
+    });
+    it('should work for 3 dimensions', () => {
+        expect(nextPageOffset({ offset: [0, 0, 0], total: [501, 501, 501] })).toEqual([0, 0, 500]);
+        expect(nextPageOffset({ offset: [0, 0, 0], total: [501, 501, 501] })).toEqual([0, 0, 500]);
```
Comment (vaclavbohac) [inline test/execution/execution-afm.test.js]: Is it exactly the same thing as the line above?

File: src/execution/execute-afm.js [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        if (newOffset[i] + PAGE_SIZE < total[i]) {
+            newOffset[i] += PAGE_SIZE;
+            return newOffset;
+        }
+        newOffset[i] = 0;
+    }
+    return false;
+}
+
+function getOnePage(pollingUri, offset, resultSoFar = false) {
```
Comment (vaclavbohac) [inline src/execution/execute-afm.js]: Would it make sense to rename the `resultSoFar` – which seemed like boolean flag to me at first to me – to `prevResult = null`?

File: test/execution/execution-afm.test.js (hunk 3 above)
Comment (zbycz) [reply test/execution/execution-afm.test.js]: oh yes, thanks for pointing out. Fixing.

File: src/execution/execute-afm.js (hunk 4 above)
Comment (zbycz) [reply src/execution/execute-afm.js]: Much better 👍

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-js-opensource-repository-check/54/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-el7/295/ : SUCCESS in 2m 05s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-docker-el7/6/ : SUCCESS in 2m 09s (non-voting)

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-js-opensource-repository-check/56/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-el7/297/ : FAILURE in 1m 31s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-docker-el7/8/ : FAILURE in 1m 44s (non-voting)

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-js-opensource-repository-check/58/ : SUCCESS in 11s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-el7/299/ : FAILURE in 1m 46s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-docker-el7/10/ : FAILURE in 1m 47s (non-voting)

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-js-opensource-repository-check/59/ : SUCCESS in 9s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-el7/300/ : UNSTABLE in 2m 03s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-docker-el7/11/ : UNSTABLE in 2m 30s (non-voting)

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-js-opensource-repository-check/60/ : SUCCESS in 7s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-el7/301/ : SUCCESS in 2m 05s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-docker-el7/12/ : SUCCESS in 1m 09s (non-voting)

**Comment by zbycz** [comment]
`gooddata@4.3.1-pavelzgdc-pzb-execute-afm-2017-11-09T15-02-20-735Z`

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-js-opensource-repository-check/61/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-el7/302/ : SUCCESS in 56s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-docker-el7/13/ : SUCCESS in 2m 07s (non-voting)

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-js-opensource-repository-check/62/ : SUCCESS in 11s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-el7/303/ : SUCCESS in 3m 14s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-docker-el7/14/ : SUCCESS in 2m 17s (non-voting)

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-js-opensource-repository-check/63/ : SUCCESS in 9s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-el7/304/ : SUCCESS in 3m 14s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-docker-el7/15/ : SUCCESS in 2m 17s (non-voting)

**Comment by zbycz** [comment]
Latest release `gooddata@4.3.1-pavelzgdc-pzb-execute-afm-2017-11-13T12-38-52-853Z`

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-js-opensource-repository-check/65/ : SUCCESS in 10s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-el7/306/ : SUCCESS in 1m 56s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-docker-el7/17/ : SUCCESS in 2m 12s (non-voting)

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-js-opensource-repository-check/66/ : SUCCESS in 9s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-el7/307/ : SUCCESS in 2m 11s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-docker-el7/18/ : SUCCESS in 2m 19s (non-voting)

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-js-opensource-repository-check/67/ : SUCCESS in 27s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-el7/308/ : SUCCESS in 2m 18s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-docker-el7/19/ : SUCCESS in 1m 10s (non-voting)

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-js-opensource-repository-check/68/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-el7/309/ : SUCCESS in 53s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-docker-el7/20/ : SUCCESS in 2m 16s (non-voting)

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://ci.intgdc.com/job/gooddata-js-merge-compliance-check/47/ : SUCCESS in 12s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-el7/310/ : SUCCESS in 52s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-docker-el7/21/ : SUCCESS in 1m 12s (non-voting)

**Pull request merged**